File: code/geometry/intersect_ll.py
def intersectLL(s1, e1, s2, e2):
	d = cross(vecsub(e1, s1), vecsub(e2, s2))
	if zero(d): # parallel
		return (-zero(cross(e1, s2, s1)), (0, 0))
	p, q = cross(e1, e2, s2), cross(e2, s1, s2)
	return (1, ((s1[0] * p + e1[0] * q) / d, (s1[1] * p + e1[1] * q) / d))

File: code/geometry_cpp/intersect_ll.cpp
pair<int, dpoint> intersectLL(ipoint s1, ipoint e1, ipoint s2, ipoint e2) {
	auto d = (e1 - s1).cross(e2 - s2);
	if (zero(d)) # parallel
		return { -int(zero(e1.cross(s2, s1))), {} };
	auto p = e1.cross(e2, s2);
	auto q = e2.cross(s1, s2);
	return { 1, dpoint(
		(s1.x * p + e1.x * q),
		(s1.y * p + e1.y * q)
	) / d };
}
